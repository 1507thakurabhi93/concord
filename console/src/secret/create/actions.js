const NAMESPACE = "user/secret";

const types = {
    USER_SECRET_CREATE_KEYPAIR: `${NAMESPACE}/keypair/request`,
    USER_SECRET_CREATE_KEYPAIR_RESPONSE: `${NAMESPACE}/keypair/response`,
};

export default types;

export const createNewKeyPair = () => ({
    type: types.USER_SECRET_CREATE_KEYPAIR
});
