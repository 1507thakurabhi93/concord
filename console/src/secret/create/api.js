import * as common from "../../api";

export const createNewKeyPair = ( name, generatePassword = false, storePassword = "" ) => {

    // Todo: Verify that query params are not added if they are null or undefined
    const query = common.queryParams({
        name,
        generatePassword
    });

    let opts = { method: "POST", credentials: "same-origin" }

    if( storePassword ) {
        const MultipartFormData = new FormData();
        MultipartFormData.append('storePassword', storePassword);

        opts.body = MultipartFormData;
    }

    return fetch(`/api/v1/secret/keypair?${query}`, opts)
        .then(response => {
            if (!response.ok) {
                throw common.defaultError(response);
            }
            return response.json();
        });
};

export const createPlainSecret = ( name, generatePassword = false, secret ) => {

    console.log( "API! CREATE PLAIN SECRET CALLED" );
    console.log(`name: ${name} , generatePassword ${generatePassword} , secret ${secret}`);

    const query = common.queryParams({
        name,
        generatePassword
    });

    const MultipartFormData = new FormData();
    MultipartFormData.append('secret', secret);

    return fetch(`/api/v1/secret/plain?${query}`, {method: "POST", credentials: "same-origin", body: MultipartFormData })
        .then( response => {
            return response.json();
        });
}

export const uploadExistingKeyPair = ( name, generatePassword = false, publicKey, privateKey, storePassword  ) => {
    
    console.log( "API! UPLOAD ExistingKeyPair" );

    const query = common.queryParams({
        name,
        generatePassword
    });

    const MultipartFormData = new FormData();
    MultipartFormData.append('public', publicKey);
    MultipartFormData.append('private', privateKey);

    if ( storePassword )
        MultipartFormData.append('storePassword', storePassword);

    return fetch(`/api/v1/secret/keypair?${query}`, {method: "POST", credentials: "same-origin", body: MultipartFormData })
        .then( response => {
            console.log( response );
        });
}

export const createUserCredentials = ( name, generatePassword = false, username, password, storePassword  ) => {
    
    console.log( "API! CreateUser Credentials " );

    const query = common.queryParams({
        name,
        generatePassword
    });

    const MultipartFormData = new FormData();
    MultipartFormData.append('username', username);
    MultipartFormData.append('password', password);

    if ( storePassword )
        MultipartFormData.append('storePassword', storePassword);

    return fetch(`/api/v1/secret/password?${query}`, {method: "POST", credentials: "same-origin", body: MultipartFormData })
        .then( response => {
            console.log( response );
        });
}

// TODO: UserCredentials

// export const uploadNewKeyPair = ( name ) => {
    
//     const query = common.queryParams({
//         name
//     });

//     return fetch(`/api/v1/secret/keypair?${query}`, { method: "POST", credentials: "same-origin"})
//         .then(response => {
//             console.log(response);
//             if (!response.ok) {
//                 throw common.defaultError(response);
//             }
//             return response.json();
//         })
//         .then(json => {
//             console.log(json);
//             return json;
//         });
// };


