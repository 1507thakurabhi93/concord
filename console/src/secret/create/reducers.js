import { combineReducers } from "redux";
import * as common from "../../reducers/common";
import types from "./actions";

const initialState = {
    power: {
        level: 100
    }
}

const test = (state=initialState, action) => {
    switch (action) {
        case types.USER_SECRET_CREATE_KEYPAIR:
            return Object.assign({}, state, {
                power: {
                    level: state.power.level + 1
                }
            });
        default:
            return state;
    }
};

export default combineReducers({ changesQuestions });

const mapStateToProps = ( state ) => ({
    power: {
        level: state.level
    }
});

const mapDispatchToProps = (dispatch) => ({
    doLogout: () => dispatch(actions.logOut())
});


