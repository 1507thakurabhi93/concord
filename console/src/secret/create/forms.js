import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Container, Button, Form, Message } from 'semantic-ui-react';
import { createNewKeyPair } from './actions';

const CreateNewKeyPair = () => {

    const handler = () => console.log("Just a secret button press");

    return (
        <Container text>
            <h2>Create New Key Pair</h2>
            <Form success>
                <Form.Input label='Secret Name' placeholder='Concord ID' />
                <Message
                    success
                    header='Created Secret'
                    content="Your public key is"
                />
                <Button onClick={handler}>Generate New Keypair</Button>
            </Form>
        </Container>
    )
}

export {
    CreateNewKeyPair
}