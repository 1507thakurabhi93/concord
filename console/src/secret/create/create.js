import { combineReducers } from "redux";
import {call, fork, put, takeLatest} from "redux-saga/effects";
import * as api from "./api";

const NAMESPACE = "user/secret";

// Actions

export const types = {
    CREATE_KEYPAIR: `${NAMESPACE}/create/keypair`,
    CREATE_WITH_EXISTING_KEYPAIR: `${NAMESPACE}/create/existingkeypair`,
    CREATE_WITH_USER_CREDENTIALS: `${NAMESPACE}/create/usercredentials`,
    CREATE_PLAIN: `${NAMESPACE}/create/simple`,

    SAVE_PUBKEY: `${NAMESPACE}/create/savepublickey`,
    START_LOADING: `${NAMESPACE}/keypair/loading`,
    CREATE_FAILED: `${NAMESPACE}/keypair/fail`,
    CREATE_SUCCESS: `${NAMESPACE}/keypair/success`
};

// Reducers 

const success = ( state: boolean = false, action ): boolean => {
    switch (action.type) {
        case types.CREATE_SUCCESS:
            return true;
        case types.CREATE_FAILED:
            return false;
        default:
            return state;
    }
}

const error = ( state: boolean = false, action ): boolean => {
    switch (action.type) {
        case types.CREATE_SUCCESS:
            return false;
        case types.CREATE_FAILED:
            return true;
        default:
            return state;
    }
}

const publicKey = ( state = null, action ) => {
    switch (action.type) {
        case types.SAVE_PUBKEY:
            return action.publicKey || null;
        case types.CREATE_PLAIN:
        case types.CREATE_WITH_USER_CREDENTIALS:
        case types.CREATE_WITH_EXISTING_KEYPAIR:
        case types.CREATE_FAILED:
            return null;
        default:
            return state;
    }
}

const message = ( state: string = '', action ): string => {
    switch (action.type) {
        case types.CREATE_SUCCESS:
            return action.message || "CREATE SUCCESS";
        case types.CREATE_FAILED:
            return action.message || "CREATE FAILED";
        default:
            return state;
    }
}

const isLoading = ( state = false, action ) => {
    switch (action.type) {
        case types.START_LOADING:
            return true
        case types.CREATE_SUCCESS:
            return false;
        case types.CREATE_FAILED:
            return false;
        default:
            return state;
    }
}

export const reducers = combineReducers({ error, isLoading, message, publicKey, success });

// Action Creators

export const actions = {
    createNewKeyPair: ( values ) => {
        return ({
            type: types.CREATE_KEYPAIR,
            payload: {
                name: values.ConcordId, 
                generatePassword: values.generatePassword
                // storePassword: values.storePassword
            }
        });
    },
    createWithExistingKeys: ( values ) => {
        return ({
            type: types.CREATE_WITH_EXISTING_KEYPAIR,
            payload: {
                name: values.ConcordId,
                publicKey: values.publicKey,
                privateKey: values.privateKey
                // generatePassword: values.generatePassword
                // storePassword: values.storePassword
            }
        });
    },
    createWithUserCredentials: ( values ) => {
        return ({
            type: types.CREATE_WITH_USER_CREDENTIALS,
            payload: {
                name: values.ConcordId,
                generatePassword: values.generatePassword,
                username: values.username,
                password: values.password
                // storePassword: values.storePassword
            }
        });
    },
    createPlainSecret: ( values ) => {
        return ({
            type: types.CREATE_PLAIN,
            payload: {
                name: values.ConcordId,
                secret: values.secret,
                generatePassword: values.generatePassword
                // storePassword: values.storePassword
            }
        });
    },
    createSuccess: ( message = "CREATE SUCCESS" ) => {
        return ({
            type: types.CREATE_SUCCESS,
            message
        });
    },
    createFailed: ( message = "CREATE FAILED" ) => {
        return ({
            type: types.CREATE_FAILED,
            message
        });
    },
    savePublicKey: ( publicKey ) => {
        return ({
            type: types.SAVE_PUBKEY,
            publicKey
        });
    },
    startLoading: () => {
        return ({
            type: types.START_LOADING
        });
    }
}

// Side Effects: Sagas

export function* createNewKeyPair( action ) {
    try {

        yield put( actions.startLoading() );

        const response = yield call( api.createNewKeyPair, action.payload );

        yield put( actions.savePublicKey( response.publicKey ) );

        yield put( actions.createSuccess("Successfully Generated a New Secret Key Pair!  Save your public Key!") );

    } catch (e) {

        yield put( actions.createFailed( "Issues encountered while creating a new KeyPair.  Name may already be taken or Data may be malformed." ) );
    
    }
}

export function* createPlainSecret( action ) {
    try {

        yield put( actions.startLoading() );

        const response = yield call( api.createPlainSecret, action.payload );

        yield put( actions.createSuccess( `Successfully Created Plain Secret: ${JSON.stringify(response)}` ) );

    } catch (e) {

        yield put( actions.createFailed( `Issues Encountered while creating new Plain Secret: ${e.message}` ) );

    }
}

export function* uploadExistingKey( action ) {
    try {

        yield put( actions.startLoading() );

        const response = yield call( api.uploadExistingKeyPair, action.payload );

        yield put( actions.createSuccess( `Successfully Uploaded Existing Keypair! ${JSON.stringify(response) || ''}` ) );

    } catch (e) {

        yield put( actions.createFailed( `Issues encountered while uploading your Keypair: ${e.message}` ) );

    }
}

export function* createUserCredentials( action ) {
    try {

        yield put( actions.startLoading() );

        const response = yield call( api.createUserCredentials, action.payload );

        yield put( actions.createSuccess( `Successfully created User Credentials Secret! ${JSON.stringify(response) || ''}` ) );

    } catch (e) {

        yield put( actions.createFailed( `Issues encountered while creating new User Credentials ${e.message}` ) );
    
    }
}

export const sagas = function*() {

    yield [
        fork(takeLatest, types.CREATE_KEYPAIR, createNewKeyPair),
        fork(takeLatest, types.CREATE_PLAIN, createPlainSecret),
        fork(takeLatest, types.CREATE_WITH_EXISTING_KEYPAIR, uploadExistingKey),
        fork(takeLatest, types.CREATE_WITH_USER_CREDENTIALS, createUserCredentials)
    ];

}
