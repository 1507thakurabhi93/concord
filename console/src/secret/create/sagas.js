// @flow
import {call, fork, put, takeLatest} from "redux-saga/effects";
import * as api from "./api";
import types from "./actions";

function* createNewKeyPair(action: any): Generator<*, *, *> {
    try {
        const response = yield call(api.createNewKeyPair);
        yield put({
            type: types.USER_SECRET_CREATE_KEYPAIR,
            response
        });
    } catch (e) {
        yield put({
            type: types.USER_SECRET_CREATE_KEYPAIR_RESPONSE,
            error: true,
            message: e.message || "Error while creating new keypair"
        });
    }
}

export default function*(): Generator<*, *, *> {
    yield [
        fork(takeLatest, types.USER_SECRET_CREATE_KEYPAIR, createNewKeyPair)
    ];
}
