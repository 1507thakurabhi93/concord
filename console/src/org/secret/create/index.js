/*-
 * *****
 * Concord
 * -----
 * Copyright (C) 2017 Wal-Mart Store, Inc.
 * -----
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =====
 */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, Header, Message, TextArea } from 'semantic-ui-react';
import { CopyToClipboard } from 'react-copy-to-clipboard';

import ErrorMessage from '../../../shared/ErrorMessage';
import { NewSecretForm } from '../components';
import * as api from './api';
import { actions, reducers, sagas, selectors } from './effects';
import { getCurrentOrg } from '../../../session/reducers';

import './styles.css';

class CreateSecretPage extends Component {
    componentDidMount() {
        this.props.resetFn();
    }

    static renderResponse({ password, publicKey }) {
        return (
            <Message success>
                <Message.Header>Secret created</Message.Header>

                {publicKey && (
                    <div>
                        <b>Public key: </b>
                        <CopyToClipboard text={publicKey}>
                            <Button icon="copy" size="mini" basic />
                        </CopyToClipboard>
                        <TextArea id="publicKeyValue" className="secretData">
                            {publicKey}
                        </TextArea>
                    </div>
                )}

                {password && (
                    <div>
                        <b>Export password: </b>
                        <CopyToClipboard text={password}>
                            <Button icon="copy" size="mini" basic />
                        </CopyToClipboard>
                        <TextArea id="passwordValue" className="secretData">
                            {password}
                        </TextArea>
                    </div>
                )}
            </Message>
        );
    }

    render() {
        const { submitFn, org, response, error, ...rest } = this.props;

        const nameCheckFn = (secretName) => {
            return api.exists(org.name, secretName).then((exists) => {
                if (exists) {
                    const err = { name: 'Already exists' };
                    throw err;
                }
            });
        };

        return (
            <div>
                <Header as="h3">New secret</Header>

                {error && <ErrorMessage message={error} />}
                {response && response.ok && CreateSecretPage.renderResponse(response)}

                <NewSecretForm
                    onSubmit={(req) => submitFn(org.name, req)}
                    nameCheckFn={nameCheckFn}
                    {...rest}
                />
            </div>
        );
    }
}

const mapStateToProps = ({ session, secretForm }) => ({
    org: getCurrentOrg(session),
    response: selectors.response(secretForm),
    error: selectors.error(secretForm),
    loading: selectors.loading(secretForm)
});

const mapDispatchToProps = (dispatch) => ({
    resetFn: () => dispatch(actions.reset()),
    submitFn: (orgName, req) => dispatch(actions.createNewSecret(orgName, req))
});

export { reducers, sagas };

export default connect(mapStateToProps, mapDispatchToProps)(CreateSecretPage);
