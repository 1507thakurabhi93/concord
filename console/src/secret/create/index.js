import React, { Component } from "react";
import {connect} from "react-redux";
import { CreateNewKeyPair, CreatePlainSecret, CreateUserCredentials, UploadExistingKeys } from "../components";
import { actions, sagas, reducers } from "./create";
import { Container, Segment, SegmentGroup, Message } from "semantic-ui-react";


export class SecretCreate extends Component {

    constructor( props ) {
        super( props );

        this.state = {
            success: false,
            error: false
        };
    }

    render() {

        const { success, error, message, publicKey, isLoading,
                Submit_NewKeyPair, 
                Submit_PlainSecret, 
                Submit_UserCredentials, 
                Submit_ExistingKeyPair,
        } = this.props;

        return (
            <Container > 
                <SegmentGroup>
                    <Segment loading={ isLoading }>
                        <CreateNewKeyPair onSubmit={ Submit_NewKeyPair } />
                        
                        <CreatePlainSecret onSubmit={ Submit_PlainSecret } />

                        <CreateUserCredentials onSubmit={ Submit_UserCredentials } />

                        <UploadExistingKeys onSubmit={ Submit_ExistingKeyPair } />

                        <Message success 
                            hidden={ !success }
                            header={ message }
                            content={ publicKey } />

                        <Message error 
                            hidden={ !error }
                            header={ message } />
                    </Segment>
                </SegmentGroup>
            </Container> 
        )
    }
}

const mapStateToProps = ( state ) => {
    return {
        success: state.secretForm.success,
        error: state.secretForm.error,
        publicKey: state.secretForm.publicKey,
        message: state.secretForm.message,
        isLoading: state.secretForm.isLoading
    };
}

const mapDispatchToProps = (dispatch) => ({
    Submit_NewKeyPair: ( values ) => dispatch( actions.CreateNewKeyPair( values ) ),
    Submit_PlainSecret: ( values ) => dispatch( actions.CreatePlainSecret( values ) ),
    Submit_UserCredentials: ( values ) => dispatch( actions.CreateWithUserCredentials( values ) ),
    Submit_ExistingKeyPair: ( values ) => dispatch( actions.CreateWithExistingKeys( values ) )
});

export default connect(mapStateToProps, mapDispatchToProps)(SecretCreate);

export { reducers, sagas }