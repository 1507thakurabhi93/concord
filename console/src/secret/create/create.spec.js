import React from 'react'
import { types, actions, reducers, sagas } from './create'

import { shallow, mount } from 'enzyme'

describe('Secret Create', () => {

    describe('reducers and actions', () => {
           
        const initialState = {
            error: false,
            success: false,
            publicKey: null,
            message: '',
            isLoading: false
        };

        it('default initial return should match initial state', () => {
            const resultState = reducers( undefined, { type: '' });
            expect( resultState ).toEqual( initialState );
        });

        it('should match snapshot for CreateSuccess Action', () => {
            const resultState = reducers( undefined, actions.createSuccess() );
            expect( resultState ).toMatchSnapshot();
        });

        it('should match snapshot for CreateFailed Action', () => {
            const resultState = reducers( undefined, actions.createFailed() );
            expect( resultState ).toMatchSnapshot();
        });

        it('should match snapshot for CreateNewKeyPair', () => {
            const resultState = reducers( undefined, actions.createNewKeyPair( "doop", false ) );
            expect ( resultState ).toMatchSnapshot();
        });

    })

})