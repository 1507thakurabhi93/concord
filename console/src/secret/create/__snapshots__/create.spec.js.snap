// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Secret Create reducers and actions should match snapshot for CreateFailed Action 1`] = `
Object {
  "error": true,
  "isLoading": false,
  "message": "CREATE FAILED",
  "publicKey": null,
  "success": false,
}
`;

exports[`Secret Create reducers and actions should match snapshot for CreateNewKeyPair 1`] = `
Object {
  "error": false,
  "isLoading": false,
  "message": "",
  "publicKey": null,
  "success": false,
}
`;

exports[`Secret Create reducers and actions should match snapshot for CreateSuccess Action 1`] = `
Object {
  "error": false,
  "isLoading": false,
  "message": "CREATE SUCCESS",
  "publicKey": null,
  "success": true,
}
`;
